urllib3/util/connection.py

```diff
@@ -57,6 +61,12 @@ def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
     # The original create_connection function always returns all records.
     family = allowed_gai_family()
 
+    # If IPv6 and selectors are available, use the Happy Eyes algorithm.
+    # (RFC 6555 https://tools.ietf.org/html/rfc6555)
+    if HAS_IPV6 and HAS_SELECT and ENABLE_HAPPY_EYEBALLS:
+        return happy_eyeballs_algorithm((host, port), timeout,
+                                        source_address, socket_options)
```

Comment:
Should all of this be brought into the same file? I knew this functionality would have a lot to it so I made it separate, if we want it all together I can do that. :)